ballista/rust/scheduler/src/scheduler_server/event_loop.rs

```diff
@@ -50,32 +47,53 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan>
     pub fn new(
         state: Arc<SchedulerState<T, U>>,
         executors_client: ExecutorsClient,
+        is_test: bool,
     ) -> Self {
         Self {
             state,
             executors_client,
+            is_test,
         }
     }
 
-    async fn offer_resources(
-        &self,
-        job_id: String,
-    ) -> Result<Option<SchedulerServerEvent>> {
-        let mut available_executors = self.state.get_available_executors_data();
+    async fn offer_resources(&self, n: u32) -> Result<Option<SchedulerServerEvent>> {
+        let mut available_executors = if self.is_test {
```

Comment:
IMO, setting flags specific for tests in application logic is usually not an anti-pattern. It makes the code harder to read and adds unnecessary runtime overhead. If I understand it correctly, the main problem is we don't have executor heartbeats in tests causing `get_alive_executors` to return an empty set? If so, I would recommend adding a helper method to set artificial executor heartbeats directly from tests so we can keep the core logic agnostics to tests.

Reply:
It's not just the heartbeat issue. For Push-based task scheduling, the schedule will launch tasks to the corresponding executors which should start up the GRPC service. It's a bit too heavy for the unit test.

Reply:
Could we use a `cfg` to achieve the same result. So something like:

```
    #[cfg(not(test))]
    pub(crate) fn get_available_executors_data(&self) -> Vec<ExecutorData> {
        let mut res = {
            let alive_executors = self.get_alive_executors_within_one_minute();
            let executors_data = self.executors_data.read();
            executors_data
                .iter()
                .filter_map(|(exec, data)| {
                    (data.available_task_slots > 0 && alive_executors.contains(exec))
                        .then(|| data.clone())
                })
                .collect::<Vec<ExecutorData>>()
        };
        res.sort_by(|a, b| Ord::cmp(&b.available_task_slots, &a.available_task_slots));
        res
    }

    #[cfg(test)]
    pub(crate) fn get_available_executors_data(&self) -> Vec<ExecutorData> {
        let mut res: Vec<ExecutorData> =
            self.executors_data.read().values().cloned().collect();
        res.sort_by(|a, b| Ord::cmp(&b.available_task_slots, &a.available_task_slots));
        res
    }
```

and then

```
        #[cfg(not(test))]
        if num_tasks > 0 {
            self.launch_tasks(&executors_delta_data, tasks_assigment)
                .await?;
        }
```

This still makes the code slightly harder to read but it removes any runtime overhead and also localizes the additional flags to two places.